text/0000-pub-sub.md

```diff
@@ -0,0 +1,171 @@
+
+## Motivation
+
+The Folio Platform consists of a set of microservices which are organized around a central proxying API Gateway. Folio modules implement these microservices through RESTful API interfaces. In order to implement complex library workflows it is necessary to distribute the work to individual modules to perform the specialized tasks.
+
+The approach described here is to provide a messaging system whereby modules publish relevant data when ready, for other modules to consume and do their part. Modules which participate in a particular workflow can subscribe to the relevant messages. The advantage of this particular solution is that it fits well with the microservices architecture and preserves the independence of the modules. Each module plays its part but remains unaware of any role played by other modules. It does not require a centralized workflow controller.
+
+## Detailed Explanation/Design
+
+### Requirements
+#### Functional Requirements
+- A business module (back-end) can define a set of events it can send (publish). It can also declare and receive (subscribe to) event of different types.
+
+- An “Event Descriptor” must be provided to define new event types.
+  1. The event definition produces a unique “Event Type”
+  2. The “Event Descriptor” must have a detailed description for the event type and payloads which can be received within events of this type.
+  3. “Event Descriptor” does not support versioning. To update an existing event declaration, a new "Event Type" must be defined with a new "Event Descriptor". However, it is recommended that the updated "Event Type" maintain the same name followed by an incremented version number suffix. The description field of the new event type should be used to explain the difference between the two related “Event Descriptors”. 
+  4. The “Event Descriptor” must define a default time-to-live (TTL) for events of this type.
+  5. The “Event Descriptor” can optionally define data structure definitions or schemas for an event payload.
+  6. The “Event Descriptor” can optionally define events as “Signed” of “Unsigned” (not in scope of MVP, but the basis for that must be provided). “Signed” means that all events of this event type must be digitally signed by the publisher.
+
+- Each instantiated event must contain at least a UUID and an "Event Type". A payload for the event is optional and can be of any type of serializable data and it is up to a publisher and a consumer to agree on the data structures they want to exchange. If an event descriptor contains data structure definitions or schemas, they can be used for validation, marshalling/unmarshalling.
+
+- It should be possible to notify a publisher that there are no subscribers for an event type it is trying to send. It could be included in a response to an event publishing call.
```

Comment:
module-x may have an optional dependency on module-y, wherein module-x publishes event A and hopes that module-y will pick it up and act.  If there is no module-y, it may be useful to provide that feedback to module-x, either to switch to different business logic, or to notify the end user through the UI

Reply:
Yes you are right, as an example, let's say SRS module parses MARC records, stores them in the DB and publishes an event that new MARCs are available, but if nobody can receive this request and update inventory's instances accordingly it means that we cannot complete the data import process and we must notify a user about that.

Reply:
Thanks Taras for that example. I'm going to try to expand upon it slightly, to improve my understanding. I'm not trying to get into the specifics of that example.

Given a `MARC records parsed` event type
And a MARC parser module is a publisher of `MARC records parsed` events
And a MARC record archiver that is a subscriber to `MARC records parsed` events
And a inventory instance updater module that is a subscriber to `MARC records parsed` events
When a `MARC records parsed` event is published
And delivered to the archiver module within the TTL
And the TTL expires before the event is delivered to the inventory instance updater module

Does the publisher receive a callback?

Reply:
This is an interesting case that isn't explicitly covered in the RFC descriptions. 

The intent is that a callback only goes out when **no** subscribers have received the event. In which case the event hasn't been consumed anywhere and it is safe to consider it dead. This would allow the publisher to enact a retry or republish strategy if desired. 

In this case though, one of the subscribers has already received the event and presumably acted upon it.  So it is no longer safe to kill the message altogether.

I suggest that the way to approach the problem is that TTL applies to the first subscriber. If any subscriber receives delivery in that time, then the message remains alive. There is no value in notifying the publisher with a callback because there is nothing much it can do with that. 

However, the pubsub module itself retains knowledge that the second subscriber did not get delivery within the TTL, and this would be captured in its history log and reflected back to the user experience. The delivery might still happen, just not within the specified TTL. The subsequent warnings would serve to highlight the issue so that the process can be adjusted.